### PR TITLE
Ensure python and node versions remain coherent

### PR DESCRIPTION
**Pull request by echarles**
Fixes https://github.com/jupyterlab/extension-cookiecutter-ts/issues/102

Get Python version from package.json

Port of https://github.com/jtpio/jupyterlab-python-file/pull/19

### CODE REVIEW

**Comment by goanpeca** [comment]
Hi @echarles, actually I suggest we leave `__init__` and `version_info` or `VERSION_INFO` in the `__init__.py` of the package. Is what any Python user will expect when importing a python package.

```python
from some_package import __version__, version_info
```

This is more or less standard (at least the `__version__` one)

**Comment by echarles** [comment]
Thx @goanpeca Makes sense. I will push something tomorrow.

**Comment by jtpio** [comment]
Also for server extensions (which is an opt-in for this cookiecutter), it makes sense to keep the `__version__`.

For `jupyterlab-python-file` (https://github.com/jtpio/jupyterlab-python-file/pull/19) it's not strictly needed since the Python package is only used to distribute the lab extension, but users are not expected to import it.

**Comment by echarles** [comment]
Hi, I have pushed a new version which just updates `_version.py` to read the version from `package.json`.
I have tested to create an extension from this branch and it works fine.

**Comment by timkpaine** [comment]
why read from package json? why not just let bump version manage both?

**Comment by timkpaine** [comment]
this is a much cleaner way to manage versions and keep js and python in sync, here is an example:

```
[bumpversion]
current_version = (0, 2, 5, 'final', 0)
commit = True
tag = False
parse = (?P<major>\d+)\,\ (?P<minor>\d+)\,\ (?P<patch>\d+)\,\ \'(?P<release>\S+)\'\,\ (?P<build>\d+)
serialize = 
	{major}, {minor}, {patch}, '{release}', {build}

[bumpversion:part:release]
optional_value = final
values = 
	alpha
	candidate
	final

[bumpversion:part:build]

[bumpversion:file:jupyterlab_templates/_version.py]

[bumpversion:file:js/package.json]
search = "version": "{current_version}"
replace = "version": "{new_version}"
serialize = 
	{major}.{minor}.{patch}-{release}.{build}
	{major}.{minor}.{patch}
```

**Comment by echarles** [comment]
Thx @timkpaine for proposing an alternative approach. I have rename this PR to `Ensure python and node versions remain coherent ` which is the goal I think we want to ensure.

I may see a risk with the bumpversion definitions where users not aware of that approach would manually change the version, and would forget to change the other version.

File: {{cookiecutter.python_name}}/{{cookiecutter.python_name}}/_version.py [hunk 1]
```diff
@@ -1,2 +1,9 @@
...
-__version__ = ".".join(map(str, version_info))
+import json
+import os.path as osp
+
+HERE = osp.abspath(osp.dirname(__file__))
+
+with open(osp.join(HERE, 'package.json')) as fid:
+    data = json.load(fid)
+
+__version__ = pdata['version']
```
Comment (jtpio) [inline {{cookiecutter.python_name}}/{{cookiecutter.python_name}}/_version.py:9]: This seems to be a typo:

```suggestion
__version__ = data['version']
```

**Comment by echarles** [comment]
@jtpio Just committed your suggestion. Curious to know what you think about the `bumpversion` alternative. My feeling so far is that a single place to set the version would mitigate the risk to diverge.

**Comment by jtpio** [comment]
CI failure looks relevant: https://github.com/jupyterlab/extension-cookiecutter-ts/runs/1285639286

About `bump2version` it sounds like a good idea as it would encourage best practices, similar to what we already have with ESLint. However there is the risk of making the cookiecutter more complicated by adding more "features" to it. It might become intimidating to newcomers who just want to get started making extensions, and think about versioning later.

But keeping the versions in sync still sounds relevant if the only purpose of having the Python package is to distribute the extension more easily.

**Comment by echarles** [comment]
> CI failure looks relevant: https://github.com/jupyterlab/extension-cookiecutter-ts/runs/1285639286

Oops turning this PR as Draft, need to double check this.

**Comment by echarles** [comment]
Spent time to fix the issue but can not find a way to get rid of the `NameError: name '__file__' is not defined`.

This is failing on `pip install -e .` when in `setup.py` the following command is invoked `version = get_version(os.path.join(name, "_version.py"))`

Tried to move the version fetch in `__init__.py` but the same issue is shown.

Googled to find a solution (e.g. https://stackoverflow.com/questions/16771894/python-nameerror-global-name-file-is-not-defined) but the proposed approaches did not work.

**Comment by goanpeca** [comment]
I can take a look @echarles

**Comment by echarles** [comment]
@goanpeca yeah I am kind-of blocked here. Would be great that you take a stab at it. I haven't digged into the `jupyter-packaging` code, maybe something to do there.

**Comment by goanpeca** [comment]
Hi @echarles, so yeah maybe jupyter_packaging is calling something in a "funny way"

Something like this works though

```python
import inspect
import json
import os.path as osp

HERE = osp.abspath(osp.dirname(inspect.getfile(lambda: None)))

with open(osp.join(HERE, 'package.json')) as fid:
    data = json.load(fid)

__version__ = data['version']

del data
del HERE
```

The extra `del` at the end, is just to reove those vars and leave only `__version__` in the module.

**Comment by echarles** [comment]
Thx @goanpeca for the workaround. Current branch contains this fix and works fine.

Let's wait a few days to get more feedback on this fix and maybe understand the cause. I ` jupyter-packaging` is involved, we should open an issue there to log and track it.

**Comment by echarles** [comment]
We have not the following error

```
$ jupyter labextension develop --overwrite .
An error occured.
FileNotFoundError: [Errno 2] No such file or directory: '/home/runner/work/extension-cookiecutter-ts/extension-cookiecutter-ts/myextension/myextension/package.json'
```

I have printed the path in `_version.py`, and indeed, `osp.abspath(osp.dirname(inspect.getfile(lambda: None)))` returns a the folder a bit too nested (should be `extension-cookiecutter-ts/myextension/package.json' and not `extension-cookiecutter-ts/myextension/myextension/package.json'`)

**Comment by telamonian** [comment]
Guys, please. No use of explicit `del` in python code without a *really* good reason.

@echarles The "issue" with `get_version(x)` is that it can't just do `import x`, since `x` is a normal system file path. So instead it effectively implements it's own light-weight import machinery using `exec`:

```python
def get_version(file, name='__version__'):
    """Get the version of the package from the given file by
    executing it and extracting the given `name`.
    """
    path = os.path.realpath(file)
    version_ns = {}
    with io.open(path, encoding="utf8") as f:
        exec(f.read(), {}, version_ns)
    return version_ns[name]
```

With respect to the `exec`call, the contents of `file` are essentially just an in-memory string, so there's no module creation and no `__file__` attr as there would be with a standard `import` statement.

So sure we could submit a PR that manually populates `__file__` (ie `version_ns = {"__file__": file}`), but that strikes me as awfully convoluted/brittle/overengineered. There are better alternatives.

**Comment by telamonian** [comment]
@echarles Effectively what you're trying to do with this PR is to treat the root `package.json` as the sole source of truth for the project's version. Which I think makes sense for the default use-case of the 3.0 cookiecutter, ie a federated (or whatever term Jason wants us to use now) extension without a serverextension, since in that case the python project is just an empty wrapper generated for the purpose of distribution.

I see two reasonable options for this:

1. drop the `get_version` call from `setup.py` and instead just copy over the logic that you're already using to load the version from `package.json` in `_version.py`

2. use @timkpaine's pretty good suggestion of `bump2version`. It's become my own go-to tool in these situations, since it's the least bad of all the terrible options

For now, let's just go with 1., since that's simplest. We can consider the inclusion of `bump2version` in the cookiecutter at a later date/PR

**Comment by echarles** [comment]
> drop the get_version call from setup.py and instead just copy over the logic that you're already using to load the version from package.json in _version.py

I have started with that but failed in the `get_version` method of `jupyter_packaging`, as you explained, prolly because of the usage of `exec(f.read(), {}, version_ns)`. I guess the workaround you propose passing `version_ns = {"__file__": file})` can be tried out.

**Comment by telamonian** [comment]
> I have started with that but failed in the `get_version` method of `jupyter_packaging`, as you explained, prolly because of the usage of `exec(f.read(), {}, version_ns)`.

No, I mean just replace these lines from the cookiecutter's `setup.py`:

https://github.com/jupyterlab/extension-cookiecutter-ts/blob/b2673213405606d0b059871f63aca48dbd6b5ad5/%7B%7Bcookiecutter.python_name%7D%7D/setup.py#L17-L18

with

```python
with open('package.json') as f:
    version = json.load(f)['version']
```

This way, `get_version` will never actually get called. This is a much better approach than trying to fake `locals()`

**Comment by goanpeca** [comment]
Well the idea if you read the thread is that the __version__ should be available on module import. Is almost a fixed convention to be able to import the version of a module that way.

**Comment by telamonian** [comment]
Yes, and it still would be this way. Why do you think it wouldn't?

**Comment by goanpeca** [comment]
Then i did not follow 🤣.

**Comment by echarles** [comment]
@telamonian I have used your trick and opened https://github.com/jupyter/jupyter-packaging/issues/53

We are green here. Converting to `ready for review`

Thx @goanpeca for interactions and help here.

**Review by telamonian — APPROVED** [review]
LGTM!

@echarles You don't have the "Allow edits by maintainers" button checked on this PR, so I ended up opening an alternative one at #108. But since you've already made the relevant changes, I'm totally fine with closing that and going forward.

The only minor improvement I made was to recursively search subdirs for `package.json`, since I think the exact default path of `myext/static/package.json` might be pretty fragile. Feel free to take said code or leave it for the purposes of this PR

**Comment by echarles** [comment]
>  "Allow edits by maintainers" 

Where is this button? Maybe a repo wide setting?

> The only minor improvement I made was to recursively search subdirs for package.json, since I think the exact default path of myext/static/package.json might be pretty fragile. Feel free to take said code or leave it for the purposes of this PR

Yeah Looking at your https://github.com/jupyterlab/extension-cookiecutter-ts/pull/108 I see you recurse in subfolders. It sounds to me like a well defined place is maybe better for the package.json to avoid divergence in code structure.

I keep this open a bit a few days and merge after if no other comments.

**Comment by telamonian** [comment]
> It sounds to me like a well defined place is maybe better for the package.json to avoid divergence in code structure.

I was also thinking that. Overall I think it would make the most sense if we structure the cookiecutter so as to guarantee that the `package.json` at the top level of the project ends up at the top level of the python package upon install. That way it's reliably available to the python code.

The only catch is that for now, at least, `package.json` is also required (I think?) at `static/package.json`, ie the base of the federated container's assets. So we'd end up with duplicate `package.json` files in the installed python package. My gut says that's a non-issue and probably a reasonable trade-off

**Comment by telamonian** [comment]
> > "Allow edits by maintainers"
> 
> Where is this button? Maybe a repo wide setting?

By default, at least, it's a per-PR setting, under the control of the PR creator. You can find the checkbox about one screen height down from the top of a PR page, on the right:

image

**Comment by blink1073** [comment]
Closing in favor of #108.  Thanks all!

**Pull request closed without merging**